app/controllers/spree/admin/reports_controller_decorator.rb

```diff
@@ -332,4 +331,16 @@ def authorized_reports
   def timestamp
     Time.zone.now.strftime("%Y%m%d")
   end
+
+  # Loads all order cycles concerned by current user, may he be producer, distributor or coordinator
+  def load_order_cycles
+    @order_cycles = OrderCycle.active_or_complete.involving_managed_distributors_of(spree_current_user)
+
+    # Avoids duplicates
+    OrderCycle.active_or_complete.involving_managed_producers_of(spree_current_user).each do |order_cycle|
+      @order_cycles << order_cycle unless @order_cycles.include?(order_cycle)
+    end
+
+    @order_cycles.sort_by!(&:orders_close_at)
+  end
```

Comment:
we should solve this as a single query. As it is, this is much less performant and harder to maintaing than asking the DB to cook the data for us (that's what the DB is for). That fact that we're removing duplicates of a result set in Ruby is a smell.

Unless there's something I haven't foreseen, we should add a new scope in the `OrderCycle` model that fetches the order cycles concerned by the current user. "Concerned" meaning that the user is allowed to see the order cycle.

I'm pretty sure @mkllnk @Matt-Yorkley or @oeoeaio can point us to where this "concerned" logic is already implemented. I guess though, that this is essentially the order cycles of the enterprises that the user has either an enterprise relationship or an enterprise role with.

Reply:
Good point. I just saw that now, there is actually a method returning all order cycles concerned by our user. I change that right now.

Reply:
It's indeed way cleaner this way !